From 9f0d632484c6f45abbd010bc384ff77f84a757b1 Mon Sep 17 00:00:00 2001
From: Tom White <tom@cloudera.com>
Date: Tue, 20 Dec 2011 13:49:53 -0800
Subject: [PATCH 021/795] MR1: MAPREDUCE-2531. org.apache.hadoop.mapred.jobcontrol.getAssignedJobID throw class cast exception.

Author: Robert Joseph Evans
Reason: Support 0.23 API in MR1
Ref: CDH-3970
(cherry picked from commit afcb7adc674c09535975552a5653d3ce3a56ec97)
(cherry picked from commit 29f7d7d2b2108267b801d50a782665d8d4078d05)
---
 .../org/apache/hadoop/mapred/jobcontrol/Job.java   |    6 +++++-
 .../hadoop/mapred/jobcontrol/TestJobControl.java   |   19 ++++++++++++++++++-
 2 files changed, 23 insertions(+), 2 deletions(-)

diff --git a/src/mapred/org/apache/hadoop/mapred/jobcontrol/Job.java b/src/mapred/org/apache/hadoop/mapred/jobcontrol/Job.java
index 1dfd80d..a84af17 100644
--- a/src/mapred/org/apache/hadoop/mapred/jobcontrol/Job.java
+++ b/src/mapred/org/apache/hadoop/mapred/jobcontrol/Job.java
@@ -63,7 +63,11 @@ public Job(JobConf conf) throws IOException {
    * mapred framework.
    */
   public JobID getAssignedJobID() {
-    return (JobID)super.getMapredJobID();
+    org.apache.hadoop.mapreduce.JobID temp = super.getMapredJobID();
+    if(temp == null) {
+      return null;
+    }
+    return JobID.downgrade(temp);
   }
 
   /**
diff --git a/src/test/org/apache/hadoop/mapred/jobcontrol/TestJobControl.java b/src/test/org/apache/hadoop/mapred/jobcontrol/TestJobControl.java
index 4c768e6..a856f51 100644
--- a/src/test/org/apache/hadoop/mapred/jobcontrol/TestJobControl.java
+++ b/src/test/org/apache/hadoop/mapred/jobcontrol/TestJobControl.java
@@ -18,12 +18,15 @@
 
 package org.apache.hadoop.mapred.jobcontrol;
 
+import static org.mockito.Mockito.*;
+
 import java.util.ArrayList;
 
 import org.apache.hadoop.conf.Configuration;
 import org.apache.hadoop.fs.FileSystem;
 import org.apache.hadoop.fs.Path;
 import org.apache.hadoop.mapred.JobConf;
+import org.apache.hadoop.mapred.JobID;
 
 /**
  * This class performs unit test for Job/JobControl classes.
@@ -191,7 +194,21 @@ public static void doJobControlTest() throws Exception {
   public void testJobControl() throws Exception {
     doJobControlTest();
   }
-    
+  
+  public void testGetAssignedJobId() throws Exception {
+    JobConf jc = new JobConf();
+    Job j = new Job(jc);
+    //Just make sure no exception is thrown
+    assertNull(j.getAssignedJobID());
+    org.apache.hadoop.mapreduce.Job mockjob = mock(org.apache.hadoop.mapreduce.Job.class);
+    org.apache.hadoop.mapreduce.JobID jid = new org.apache.hadoop.mapreduce.JobID("test",0);
+    when(mockjob.getJobID()).thenReturn(jid);
+    j.setJob(mockjob);
+    JobID expected = new JobID("test",0);
+    assertEquals(expected, j.getAssignedJobID());
+    verify(mockjob).getJobID();
+  }
+  
   public static void main(String[] args) {
     TestJobControl test = new TestJobControl();
     try {
-- 
1.7.0.4

